readthedocs/projects/tasks/builds.py

```diff
@@ -513,48 +513,88 @@ def on_failure(self, exc, task_id, args, kwargs, einfo):
         # Update build object
         self.data.build['success'] = False
 
+    def get_valid_artifact_types(self):
+        """
+        Return a list of all the valid artifact types for this build.
+
+        It performs the following checks on each output format type path:
+         - it exists
+         - it is a directory
+         - does not contains more than 1 files (only PDF, HTMLZip, ePUB)
+
+        TODO: remove the limitation of only 1 file.
+        Add support for multiple PDF files in the output directory and
+        grab them by using glob syntaxt between other files that could be garbage.
+        """
+        for artifact_type in ARTIFACT_TYPES:
+            artifact_directory = self.data.project.artifact_path(
+                version=self.data.version.slug,
+                type_=artifact_type,
+            )
+            if not os.path.exists(artifact_directory):
+                # There is no output directory.
+                # Skip this format.
+                continue
+
+            if not os.path.isdir(artifact_directory):
+                log.exception(
+                    "The output path is not a directory.",
+                    output_format=artifact_type,
+                )
+                # TODO: we should raise an exception here, fail the build,
+                # and communicate the error to the user.
+                # We are just skipping this output format for now.
+                #
+                # raise BuildUserError(BuildUserError.BUILD_OUTPUT_IS_NOT_A_DIRECTORY.format(artifact_type))
+                continue
+
+            # Check if there are multiple files on artifact directories.
+            # These output format does not support multiple files yet.
+            # In case multiple files are found, the upload for this format is not performed.
+            #
+            # TODO: we should fail the build for these cases and clearly communicate this.
+            # to the user. To do this, we need to call this method (``store_build_artifacts``)
+            # since the ``execute`` method.
+            # It will allow us to just `raise BuildUserError` and handle it at
+            # Celery `on_failure` handler.
```

Comment:
```suggestion
            # Check if there are multiple files in artifact directories.
            # These output formats don't support multiple files yet.
            # In case multiple files are found, the upload for this format is not performed.
            #
            # TODO: we should fail the build for these cases and clearly communicate this
            # to the user. To do this, we need to call this method (``store_build_artifacts``)
            # It will allow us to just `raise BuildUserError` and handle it at
            # in the `on_failure` handler.
```

Wasn't really sure what the second part here was saying with the `since the ``execute`` method.`?

Reply:
Meh, `s/since/from inside/`. So, the correct sentence there should be:

> we should fail the build for these cases and clearly communicate this to the user. To do this, we need to call this method (``store_build_artifacts``) from inside the ``execute`` method.

Reply:
This comment was removed in the next iteration anyways, since we are calling `store_build_artifacts` from inside `execute` now and we can just raise exceptions there.